lib/SIL/Utils/PrunedLiveness.cpp

```diff
@@ -489,8 +489,50 @@ bool PrunedLiveRange<LivenessWithDefs>::isWithinBoundary(
   llvm_unreachable("instruction must be in its parent block");
 }
 
-static bool checkDeadEnd(SILInstruction *inst, DeadEndBlocks *deadEndBlocks) {
-  return deadEndBlocks && deadEndBlocks->isDeadEnd(inst->getParent());
+/// Whether \p parent is a dead (reported to be dead by `liveBlocks`), dead-end
+/// (such as an infinite loop) block within the availability boundary (where
+/// the value has not been consumed).
+static bool checkDeadEnd(SILBasicBlock *parent, DeadEndBlocks *deadEndBlocks,
+                         PrunedLiveBlocks const &liveBlocks) {
+  if (!deadEndBlocks) {
+    return false;
+  }
+  if (!deadEndBlocks->isDeadEnd(parent)) {
+    return false;
+  }
+  if (liveBlocks.getBlockLiveness(parent) != PrunedLiveBlocks::Dead) {
+    return false;
+  }
+  // Check whether the value is available in `parent` (i.e. not consumed on any
+  // path to it):
+  //
+  // Search backward until LiveOut or LiveWithin blocks are reached.
+  // (1) If ALL the reached blocks are LiveOut, then `parent` IS within the
+  //     boundary.
+  // (2) If ANY reached block is LiveWithin, then `parent` is NOT within the
```

Comment:
Do you think it's obvious enough that `LiveWithin` implies that this path reached a consume, and therefore makes the value unavailable?

Reply:
Expanded the comment to
```
(2) If ANY reached block is LiveWithin, the value was consumed in that
    reached block, preventing the value from being available at `parent`,
    so `parent` is NOT within the availability boundary.
```
and added backreferences to (1) and (2) from the switch below.